link/trace_event.go

```diff
@@ -0,0 +1,281 @@
+	}
+
+	fd, err := unix.PerfEventOpen(&attr, -1, 0, -1, unix.PERF_FLAG_FD_CLOEXEC)
+	if err != nil {
+		return 0, err
+	}
+
+	return fd, nil
+}
+
+// unsafeStringPtr returns an unsafe.Pointer to a NUL-terminated copy of str.
+func unsafeStringPtr(str string) (unsafe.Pointer, error) {
+	p, err := unix.BytePtrFromString(str)
+	if err != nil {
+		return nil, err
+	}
+	return unsafe.Pointer(p), nil
+}
+
+// getTraceEventID reads a trace event's ID from tracefs given its group and name.
+// group and name must be alphanumeric (with _-), as required by the kernel.
+func getTraceEventID(group, name string) (uint64, error) {
+	if !rgxTraceEvent.MatchString(group) {
+		return 0, fmt.Errorf("trace event group must be alphanumeric (with _-): %s", group)
+	}
+	if !rgxTraceEvent.MatchString(name) {
+		return 0, fmt.Errorf("trace event name must be alphanumeric (with _-): %s", name)
+	}
+
+	path := filepath.Join(tracefsPath, "events", group, name, "id")
+	data, err := os.ReadFile(path)
+	if os.IsNotExist(err) {
+		return 0, fmt.Errorf("trace event %s/%s: %w", group, name, ErrNotSupported)
+	}
+	if err != nil {
+		return 0, fmt.Errorf("reading trace event ID of %s/%s: %w", group, name, err)
+	}
+
+	tid := bytes.TrimSpace(data)
+	return strconv.ParseUint(string(tid), 10, 64)
+}
+
+// getPMUEventType reads a Performance Monitoring Unit's type (numeric identifier)
+// from /sys/bus/event_source/devices/<pmu>/type.
+func getPMUEventType(pmu string) (uint64, error) {
+	path := filepath.Join("/sys/bus/event_source/devices", pmu, "type")
+	data, err := os.ReadFile(path)
+	if os.IsNotExist(err) {
+		return 0, fmt.Errorf("pmu type %s: %w", pmu, ErrNotSupported)
+	}
+	if err != nil {
+		return 0, fmt.Errorf("reading pmu type %s: %w", pmu, err)
+	}
+	et := bytes.TrimSpace(data)
+	return strconv.ParseUint(string(et), 10, 64)
+}
+
+// A perfEvent represents a perf event kernel object. Exactly one eBPF program
+// can be attached to it. It is created based on a tracefs trace event or a
+// Performance Monitoring Unit (PMU).
+type perfEvent struct {
+	fd       *internal.FD
+	progType ebpf.ProgramType
+	prog     *ebpf.Program
+}
+
+func (ev *perfEvent) isLink() {}
+
+func (ev *perfEvent) Pin(string) error {
+	return fmt.Errorf("pin perf event: %w", ErrNotSupported)
+}
+
+func (ev *perfEvent) Update(prog *ebpf.Program) error {
+	if t := prog.Type(); t != ev.progType {
+		return fmt.Errorf("invalid program type (expected %s): %s", ev.progType, t)
+	}
+	if prog.FD() < 0 {
+		return fmt.Errorf("invalid program: %w", internal.ErrClosedFd)
+	}
+
+	// Return if the given prog is already attached.
+	if ev.prog == prog {
+		return nil
+	}
+
+	pfd, err := ev.fd.Value()
+	if err != nil {
+		return fmt.Errorf("getting perf event fd: %w", err)
+	}
+
+	err = unix.IoctlSetInt(int(pfd), unix.PERF_EVENT_IOC_SET_BPF, prog.FD())
+	if err != nil {
+		return fmt.Errorf("setting perf event bpf program: %w", err)
+	}
+
+	// Store a reference to the attached program.
+	ev.prog = prog
```

Comment:
True, we don't really need this. The program being closed or not doesn't really impact this particular logic, though. If the pointers match, an attach is skipped.

Now I think of it, `Attach()` should probably check if the program is still open/valid, perhaps through a `closed bool` exposed through a `(Program).Closed() bool`? Would that be useful?

Reply:
> If the pointers match, an attach is skipped.

I can call `prog.Clone()` and invoke this function. Should it bail out early or not? IMO, in the absence of a strong reason to do this ww should drop this.

> Now I think of it, Attach() should probably check if the program is still open/valid, perhaps through a closed bool exposed through a (Program).Closed() bool? Would that be useful?

Using an object after calling `Close()` on it is in the same realm as `new(Type)` for me: not worth defending against.

Reply:
If I understand the following kernel code correctly: https://elixir.bootlin.com/linux/v5.11.10/source/kernel/trace/bpf_trace.c#L1943, `PERF_EVENT_IOC_SET_BPF` doesn't look like it replaces an attached program at all. It adds it to a prog array owned by the perf event, so `Update()` will keep appending to it. Will have to play around with this a bit to see if that's actually what happens.

Reply:
I think your reading is correct, for a replace the second argument to bpf_prog_array_copy should be non-NULL.

Reply:
I'll also put these findings in the code, but calling `PERF_EVENT_IOC_SET_BPF` multiple times on the same perf event has only been possible since 4.15 (e87c6bc3852b "bpf: permit multiple bpf attachments for a single perf event"). Before that, it returns EEXIST. There doesn't seem to be a way to remove a BPF prog from a perf event (tried `PERF_EVENT_IOC_RESET` and `_DISABLE`).

The `perf_event_detach_bpf_prog` kernel fn only seems to be called during perf event teardown, and I can't find a way to trigger it using a syscall/ioctl. I'll make `Update()` not implemented for perf events for now. For kernels after 4.15, we could support attaching multiple progs to a single perf event, but the current semantics of  `Link` don't make that very intuitive. Omitting it for now would be the safest choice.